# [LAB-05] 1. Boston Housing Dataset

## #01. 준비작업

### 1. 필요한 라이브러리 참조

In [1]:
import numpy as np
from jussam import load_data
from pandas import DataFrame, read_excel

📦 연세대학교 주영아 교수가 제작한 라이브러리를 사용중입니다.
📧 Email(1): j.purplerose@yonsei.ac.kr
📧 Email(2): j.purplerose@gmail.com
📝 Website: https://juyounga.kr/


### 2. 데이터 불러오기

In [2]:
origin = load_data("boston")
origin.head()

📚 보스턴 인근 지역의 인구·환경·교육·범죄·세금 정보 등을 기반으로 주택의 중위가격(MEDV)을 예측하기 위해 수집된 대표적인 회귀 분석용 주택시장 데이터 (출처: https://www.kaggle.com/datasets/altavish/boston-housing-dataset)

    field    description
--  -------  ---------------------------------------------------------------------------------------------
 0  CRIM     타운별 1인당 범죄율 (Per capita crime rate by town)
 1  ZN       25,000 sqft 이상 주거용 대지 비율 (Proportion of residential land zoned for large lots)
 2  INDUS    비소매 상업구역 비율 (Proportion of non-retail business acres per town)
 3  CHAS     명목형, 찰스강 인접 여부 (Charles River dummy variable: 1 if tract bounds river; 0 otherwise)
 4  NOX      대기 오염 정도 (Nitric oxides concentration, parts per 10 million)
 5  RM       주택당 평균 방(Room) 수 (Average number of rooms per dwelling)
 6  AGE      1940년 이전 건축 비율 (Proportion of owner-occupied units built prior to 1940)
 7  DIS      5개 주요 고용센터까지의 거리 (Weighted distances to five Boston employment centers)
 8  RAD      고속도로 접근성 지수 (Index of accessibility to radial highways)
 9  T

,CRIM,ZN,INDUS,CHAS,NOX,RM,AGE,DIS,RAD,TAX,PTRATIO,B,LSTAT,MEDV
0,0.006,18.000,2.310,0,0.538,6.575,65.200,4.090,1,296,15.300,396.900,4.980,24.000
1,0.027,0.000,7.070,0,0.469,6.421,78.900,4.967,2,242,17.800,396.900,9.140,21.600
2,0.027,0.000,7.070,0,0.469,7.185,61.100,4.967,2,242,17.800,392.830,4.030,34.700
3,0.032,0.000,2.180,0,0.458,6.998,45.800,6.062,3,222,18.700,394.630,2.940,33.400
4,0.069,0.000,2.180,0,0.458,7.147,54.200,6.062,3,222,18.700,396.900,5.330,36.200


## #02. 자료형 점검

### 1. 자료형 확인

In [3]:
origin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    int64  
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    int64  
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(11), int64(3)
memory usage: 55.5 KB


### 2. 자료형 변환

- CHAS는 범주형이므로 category 타입으로 변환

In [4]:
# 원본 데이터 프레임 복사
df1 = origin.copy()
df1["CHAS"] = df1["CHAS"].astype("category")
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype   
---  ------   --------------  -----   
 0   CRIM     506 non-null    float64 
 1   ZN       506 non-null    float64 
 2   INDUS    506 non-null    float64 
 3   CHAS     506 non-null    category
 4   NOX      506 non-null    float64 
 5   RM       506 non-null    float64 
 6   AGE      506 non-null    float64 
 7   DIS      506 non-null    float64 
 8   RAD      506 non-null    int64   
 9   TAX      506 non-null    int64   
 10  PTRATIO  506 non-null    float64 
 11  B        506 non-null    float64 
 12  LSTAT    506 non-null    float64 
 13  MEDV     506 non-null    float64 
dtypes: category(1), float64(11), int64(2)
memory usage: 52.1 KB


## #03. 데이터 중복 점검

### 1. 행단위 완전 중복 확인

- 중복된 행에 대하여 True로 표시하고, 중복되지 않은 행은 False로 표시한다.

In [5]:
dup = df1.duplicated()
dup

0      False
1      False
2      False
3      False
4      False
       ...  
501    False
502    False
503    False
504    False
505    False
Length: 506, dtype: bool

In [6]:
# 중복 데이터 개수
dup.sum()

np.int64(0)

### 2. 중복 데이터 제거

- 여기서는 필요 없는 처리이지만 코드 학습을 위해 수행

In [7]:
df2 = df1.drop_duplicates()
df2.duplicated().sum()

np.int64(0)

## #03. 형식·표기·단위 점검

### 1. 명목형 변수의 단위 확인

- 명목형 변수는 고유값의 분포를 확인하는 것으로 단위 점검을 수행할 수 있다.

In [8]:
df2['CHAS'].value_counts()

CHAS
0    471
1     35
Name: count, dtype: int64

> 이진 분류이고, 이미 라벨링이 되어 있으므로 추가적인 처리(라벨링, 더미변수)는 필요 없다고 판단

### 2. 연속형 변수의 단위 확인을 위한 변수명 추출

In [9]:
# 값을 검사 수치형 변수의 이름 추출
fields = df2.select_dtypes(include="number").columns.to_list()
print(fields)

['CRIM', 'ZN', 'INDUS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']


### 3. 연속형 변수의 단위 확인

- 값의 구간(최소,최대값)을 확인하여 데이터 명세와 비교해 본다.

In [10]:
minmax = [] # 최소값과 최대값을 저장할 리스트

for field in fields:
    min_value = df2[field].min()
    max_value = df2[field].max()
    minmax.append({"min": min_value, "max": max_value})

# 최소값과 최대값을 데이터 프레임으로 변환하여 출력
minmax_df = DataFrame(minmax, index=fields)
minmax_df

,min,max
CRIM,0.006,88.976
ZN,0.000,100.000
INDUS,0.460,27.740
NOX,0.385,0.871
RM,3.561,8.780
AGE,2.900,100.000
DIS,1.130,12.127
RAD,1.000,24.000
TAX,187.000,711.000
PTRATIO,12.600,22.000


## #05. 결측치 점검

### 1. 결측치 확인

- 데이터프레임 객체(df1)의 isna() 메서드는 결측치인 경우 True, 결측치가 아닌 경우 False로 채워진 데이터프레임을 반환한다.
- 결측치 여부를 담고 있는 데이터프레임을 sum() 메서드로 집계하면 True를 1, False를 0으로 환산하여 열 단위로 집계하는데 이는 열 단위의 결측치 수를 의미한다.

In [11]:
na_count = df2.isna().sum()
na_count

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64

### 2. 결측치 비율 계산을 위한 데이터 크기 확인

- 이미 앞 과정에서 결측치 없음이 확인되었지만 결측치가 있는 상황을 가정해서 코드 실습을 위해 진행함

In [12]:
# 데이터의 행과 열의 수 확인
rows, cols = df2.shape
print(f"rows: {rows}, cols: {cols}")

rows: 506, cols: 14


### 3. 결측치 비율 확인

- 각 변수의 결측치 수를 전체 행 수로 나눈다.
- 결측치 비율이 5%(0.05) 미만이므로 전체 결과에 미치는 영향이 미미 하므로 결측치 삭제 가능 (Ryan, 2013)

In [13]:
na_ratio = na_count / rows
na_ratio

CRIM      0.000
ZN        0.000
INDUS     0.000
CHAS      0.000
NOX       0.000
RM        0.000
AGE       0.000
DIS       0.000
RAD       0.000
TAX       0.000
PTRATIO   0.000
B         0.000
LSTAT     0.000
MEDV      0.000
dtype: float64

## #06. 품질 점검 결과 데이터셋 저장

- 다음 분석 과정에서 이어서 진행할 수 있도록 품질 점검이 완료된 데이터 셋을 저장한다.

In [14]:
df2.to_excel("boston_qtcheck.xlsx", index=False)

## #07. 기술 통계량 확인 준비작업

### 1. 데이터 불러오기

- 실제 분석 과정에서는 새로운 jupyter 파일을 생성하고 앞 단계에서 저장한 데이터를 불러와서 진행하는 것을 권장

In [15]:
#origin_qt = read_excel("boston_qtcheck.xlsx")

# 수업용 자료이므로 라이브러리를 통해서도 로드 가능함
origin_qt = load_data("boston_qtcheck")

📚 보스턴 주택 가격 데이터셋의 품질 검사 완료 버전 (출처: 자체 정제)

    field    description
--  -------  ---------------------------------------------------------------------------------------------
 0  CRIM     타운별 1인당 범죄율 (Per capita crime rate by town)
 1  ZN       25,000 sqft 이상 주거용 대지 비율 (Proportion of residential land zoned for large lots)
 2  INDUS    비소매 상업구역 비율 (Proportion of non-retail business acres per town)
 3  CHAS     명목형, 찰스강 인접 여부 (Charles River dummy variable: 1 if tract bounds river; 0 otherwise)
 4  NOX      대기 오염 정도 (Nitric oxides concentration, parts per 10 million)
 5  RM       주택당 평균 방(Room) 수 (Average number of rooms per dwelling)
 6  AGE      1940년 이전 건축 비율 (Proportion of owner-occupied units built prior to 1940)
 7  DIS      5개 주요 고용센터까지의 거리 (Weighted distances to five Boston employment centers)
 8  RAD      고속도로 접근성 지수 (Index of accessibility to radial highways)
 9  TAX      재산세율 (Full-value property-tax rate per $10,000)
10  PTRATIO  교육구 학생-교사 비율 (Pupil–teacher ratio by town)
11  

### 2. 타입변환

- category 타입으로의 변환은 데이터를 불러올 때 마다 진행해야 한다.

In [16]:
df3 = origin_qt.copy()
df3['CHAS'] = df3['CHAS'].astype('category')
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype   
---  ------   --------------  -----   
 0   CRIM     506 non-null    float64 
 1   ZN       506 non-null    float64 
 2   INDUS    506 non-null    float64 
 3   CHAS     506 non-null    category
 4   NOX      506 non-null    float64 
 5   RM       506 non-null    float64 
 6   AGE      506 non-null    float64 
 7   DIS      506 non-null    float64 
 8   RAD      506 non-null    int64   
 9   TAX      506 non-null    int64   
 10  PTRATIO  506 non-null    float64 
 11  B        506 non-null    float64 
 12  LSTAT    506 non-null    float64 
 13  MEDV     506 non-null    float64 
dtypes: category(1), float64(11), int64(2)
memory usage: 52.1 KB


## #08. 기술 통계량 표 생성

### 1. 연속형 변수의 기술 통계량 표

- 명목형 변수는 자동으로 제외된다.

In [17]:
# T는 데이터프레임의 행과 열을 바꿔주는 역할
desc_df = df3.describe().T
desc_df

,count,mean,std,min,25%,50%,75%,max
CRIM,506.000,3.614,8.602,0.006,0.082,0.257,3.677,88.976
ZN,506.000,11.364,23.322,0.000,0.000,0.000,12.500,100.000
INDUS,506.000,11.137,6.860,0.460,5.190,9.690,18.100,27.740
NOX,506.000,0.555,0.116,0.385,0.449,0.538,0.624,0.871
RM,506.000,6.285,0.703,3.561,5.885,6.208,6.623,8.780
AGE,506.000,68.575,28.149,2.900,45.025,77.500,94.075,100.000
DIS,506.000,3.795,2.106,1.130,2.100,3.207,5.188,12.127
RAD,506.000,9.549,8.707,1.000,4.000,5.000,24.000,24.000
TAX,506.000,408.237,168.537,187.000,279.000,330.000,666.000,711.000
PTRATIO,506.000,18.456,2.165,12.600,17.400,19.050,20.200,22.000


### 2. 명목형 변수의 기술 통계량 표

- 데이터의 수(count), 값의 종류 수(unique), 최빈값(top), 최빈값의 수(freq)를 알 수 있다.

In [18]:
cate_desc_df = df3.describe(include='category').T
cate_desc_df

,count,unique,top,freq
CHAS,506,2,0,471


### 3. 명목형 변수 기술 통계량

- 명목형 변수는 값의 종류별로 데이터 수와 비율을 확인할 수 있다.

In [19]:
# 명목형 변수의 필드명 추출
cate_fields = df3.select_dtypes(include='category').columns

# 명목형 변수의 고유값과 빈도수 계산
for field in cate_fields:
    # 값의 종류별 개수
    vcount = df3[field].value_counts()
    # 전체 데이터 길이 대비 비율
    percent = vcount / df3.shape[0]

    # 결과 저장
    df = DataFrame({'count': vcount, 'percent': percent})
    # 결과 출력
    display(df)

,count,percent
CHAS,,
0,471,0.931
1,35,0.069


## #09. 중심 수준 파악

### 1. **평균-중앙값의 상대 차이율** 계산

- 평균-중앙값 상대 차이율 = |평균 - 중앙값| / 중앙값

| 구간 | 의미 |
|--------|-----------------|
| 0 ~ 0.1   | 평균과 중앙값이 거의 비슷함 |
| 0.1 ~ 0.5  | 약간 차이 있음        |
| 0.5 | 차이 큼, 왜도·극단값 의심 |

In [20]:
# "평균-중앙값 상대 차이율 = |평균 - 중앙값| / 중앙값" 컬럼 추가
desc_df['rel_diff'] = abs(desc_df['mean'] - desc_df['50%']) / desc_df['50%']

# 상대 차이율 의미 컬럼 추가
conditions = [desc_df['rel_diff'] < 0.1, desc_df['rel_diff'] < 0.5]
choices = ['similar', 'diff']
desc_df['rdiff_flag'] = np.select(conditions, choices, default='large_diff')

# 필요한 컬럼만 선택하여 출력
desc_df[['mean', '50%', 'rel_diff', 'rdiff_flag']]

,mean,50%,rel_diff,rdiff_flag
CRIM,3.614,0.257,13.087,large_diff
ZN,11.364,0.000,inf,large_diff
INDUS,11.137,9.690,0.149,diff
NOX,0.555,0.538,0.031,similar
RM,6.285,6.208,0.012,similar
AGE,68.575,77.500,0.115,diff
DIS,3.795,3.207,0.183,diff
RAD,9.549,5.000,0.910,large_diff
TAX,408.237,330.000,0.237,diff
PTRATIO,18.456,19.050,0.031,similar


## #10. 이상치 신호 탐지

### 1. IQR, 이상치 경계값 계산

In [21]:
# iqr
desc_df['iqr'] = desc_df['75%'] - desc_df['25%']

# 상한 이상치 경계
desc_df['upper_bound'] = desc_df['75%'] + 1.5 * desc_df['iqr']

# 하한 이상치 경계
desc_df['lower_bound'] = desc_df['25%'] - 1.5 * desc_df['iqr']

# 필요한 컬럼만 선택하여 출력
desc_df[['iqr', 'upper_bound', 'lower_bound']]

,iqr,upper_bound,lower_bound
CRIM,3.595,9.070,-5.311
ZN,12.500,31.250,-18.750
INDUS,12.910,37.465,-14.175
NOX,0.175,0.886,0.187
RM,0.738,7.731,4.778
AGE,49.050,167.650,-28.550
DIS,3.088,9.821,-2.532
RAD,20.000,54.000,-26.000
TAX,387.000,1246.500,-301.500
PTRATIO,2.800,24.400,13.200


### 2. 명목형 변수를 제외한 데이터 프레임

- 이상치는 연속형 변수에 대한 개념이므로 명목형 변수를 제외한 데이터프레임 생성

In [22]:
# 명목형 변수의 필드명 추출
cate_fields = df3.select_dtypes(include='category').columns

df4 = df3.drop(columns=cate_fields)
df4.head()

,CRIM,ZN,INDUS,NOX,RM,AGE,DIS,RAD,TAX,PTRATIO,B,LSTAT,MEDV
0,0.006,18.000,2.310,0.538,6.575,65.200,4.090,1,296,15.300,396.900,4.980,24.000
1,0.027,0.000,7.070,0.469,6.421,78.900,4.967,2,242,17.800,396.900,9.140,21.600
2,0.027,0.000,7.070,0.469,7.185,61.100,4.967,2,242,17.800,392.830,4.030,34.700
3,0.032,0.000,2.180,0.458,6.998,45.800,6.062,3,222,18.700,394.630,2.940,33.400
4,0.069,0.000,2.180,0.458,7.147,54.200,6.062,3,222,18.700,396.900,5.330,36.200


### 3. 상한 이상치 탐지

In [23]:
# 상한 이상치 수
desc_df['upper_outliers'] = ((df4 > desc_df['upper_bound'])).sum()

# 상한 이상치 수 비율
desc_df['upper_outliers_ratio'] = desc_df['upper_outliers'] / df4.shape[0]

# 필요한 컬럼만 선택하여 출력
desc_df[['upper_outliers', 'upper_outliers_ratio']]

,upper_outliers,upper_outliers_ratio
CRIM,66,0.130
ZN,68,0.134
INDUS,0,0.000
NOX,0,0.000
RM,22,0.043
AGE,0,0.000
DIS,5,0.010
RAD,0,0.000
TAX,0,0.000
PTRATIO,0,0.000


### 4. 하한 이상치 탐지

In [24]:
# 하한 이상치 수
desc_df['lower_outliers'] = (df4 < desc_df['lower_bound']).sum()

# 하한 이상치 수 비율
desc_df['lower_outliers_ratio'] = desc_df['lower_outliers'] / df4.shape[0]

# 필요한 컬럼만 선택하여 출력
desc_df[['lower_outliers', 'lower_outliers_ratio']]

,lower_outliers,lower_outliers_ratio
CRIM,0,0.000
ZN,0,0.000
INDUS,0,0.000
NOX,0,0.000
RM,8,0.016
AGE,0,0.000
DIS,0,0.000
RAD,0,0.000
TAX,0,0.000
PTRATIO,15,0.030


### 5. 전체 이상치 집계

In [25]:
# 통합 이상치 수
desc_df['outliers'] = desc_df['upper_outliers'] + desc_df['lower_outliers']

# 통합 이상치 수 비율
desc_df['outliers_ratio'] = desc_df['outliers'] / df3.shape[0]

# 이상치 탐지 결과 확인
desc_df[['upper_outliers', 'upper_outliers_ratio', 
         'lower_outliers', 'lower_outliers_ratio', 
         'outliers', 'outliers_ratio']]

,upper_outliers,upper_outliers_ratio,lower_outliers,lower_outliers_ratio,outliers,outliers_ratio
CRIM,66,0.130,0,0.000,66,0.130
ZN,68,0.134,0,0.000,68,0.134
INDUS,0,0.000,0,0.000,0,0.000
NOX,0,0.000,0,0.000,0,0.000
RM,22,0.043,8,0.016,30,0.059
AGE,0,0.000,0,0.000,0,0.000
DIS,5,0.010,0,0.000,5,0.010
RAD,0,0.000,0,0.000,0,0.000
TAX,0,0.000,0,0.000,0,0.000
PTRATIO,0,0.000,15,0.030,15,0.030


## #11. 비대칭 신호 확인

### 1. 왜도 점검

In [26]:
# 왜도 계산
desc_df['skew'] = df4.skew()

# 왜도를 통한 분포 형태 해석
conditions_skew = [(desc_df['skew'] < -0.5), (desc_df['skew'] > 0.5)]
choices_skew = ['left tail', 'right tail']
desc_df['skew_interpret'] = np.select(conditions_skew, choices_skew,
                                    default='symmetric')

# 필요한 컬럼만 선택하여 출력
desc_df[['skew', 'skew_interpret']]

,skew,skew_interpret
CRIM,5.223,right tail
ZN,2.226,right tail
INDUS,0.295,symmetric
NOX,0.729,right tail
RM,0.404,symmetric
AGE,-0.599,left tail
DIS,1.012,right tail
RAD,1.005,right tail
TAX,0.670,right tail
PTRATIO,-0.802,left tail


### 2. 첨도 점검

In [27]:
# 첨도 계산
desc_df['kurt'] = df4.kurt()

# 첨도를 통한 분포 형태 해석
conditions_kurt = [(desc_df['kurt'] < 0), (desc_df['kurt'] > 0)]
choices_kurt = ['platykurtic', 'leptokurtic']
desc_df['kurt_interpret'] = np.select(conditions_kurt, choices_kurt,
                                    default='mesokurtic')

# 필요한 컬럼만 선택하여 출력
desc_df[['kurt', 'kurt_interpret']]

,kurt,kurt_interpret
CRIM,37.131,leptokurtic
ZN,4.032,leptokurtic
INDUS,-1.234,platykurtic
NOX,-0.065,platykurtic
RM,1.892,leptokurtic
AGE,-0.968,platykurtic
DIS,0.488,leptokurtic
RAD,-0.867,platykurtic
TAX,-1.142,platykurtic
PTRATIO,-0.285,platykurtic


### 3. 로그 변환 필요성 판단 함수 정의

In [28]:
def judge_log_transform(skew, kurt):
    if skew >= 1:                        # 강한 우측 꼬리 분포
        return "log1p"
    elif skew > 0.5 and kurt > 0:       # 우측 꼬리 분포이면서 첨도가 높은 경우
        return "log1p"
    elif skew <= -1:                    # 강한 좌측 꼬리 분포
        return "reverse_log1p"
    elif skew < -0.5 and kurt > 0:      # 좌측 꼬리 분포이면서 첨도가 높은 경우
        return "reverse_log1p"
    else:                               # 대칭 분포
        return "none"

### 4. 로그 변환 필요성 판정

In [29]:
desc_df['log_need'] = desc_df.apply(lambda row: judge_log_transform( row['skew'], row['kurt']), axis=1)

# 로그 변환 필요성 결과 확인
desc_df[['skew', 'kurt', 'log_need']]

,skew,kurt,log_need
CRIM,5.223,37.131,log1p
ZN,2.226,4.032,log1p
INDUS,0.295,-1.234,none
NOX,0.729,-0.065,none
RM,0.404,1.892,none
AGE,-0.599,-0.968,none
DIS,1.012,0.488,log1p
RAD,1.005,-0.867,log1p
TAX,0.670,-1.142,none
PTRATIO,-0.802,-0.285,none


## #12. 최종 기술 통계량 확인

### 1. 최종 통계량 표 확인

In [30]:
desc_df.T

,CRIM,ZN,INDUS,NOX,RM,AGE,DIS,RAD,TAX,PTRATIO,B,LSTAT,MEDV
count,506.000,506.000,506.000,506.000,506.000,506.000,506.000,506.000,506.000,506.000,506.000,506.000,506.000
mean,3.614,11.364,11.137,0.555,6.285,68.575,3.795,9.549,408.237,18.456,356.674,12.653,22.533
std,8.602,23.322,6.860,0.116,0.703,28.149,2.106,8.707,168.537,2.165,91.295,7.141,9.197
min,0.006,0.000,0.460,0.385,3.561,2.900,1.130,1.000,187.000,12.600,0.320,1.730,5.000
25%,0.082,0.000,5.190,0.449,5.885,45.025,2.100,4.000,279.000,17.400,375.377,6.950,17.025
50%,0.257,0.000,9.690,0.538,6.208,77.500,3.207,5.000,330.000,19.050,391.440,11.360,21.200
75%,3.677,12.500,18.100,0.624,6.623,94.075,5.188,24.000,666.000,20.200,396.225,16.955,25.000
max,88.976,100.000,27.740,0.871,8.780,100.000,12.127,24.000,711.000,22.000,396.900,37.970,50.000
rel_diff,13.087,inf,0.149,0.031,0.012,0.115,0.183,0.910,0.237,0.031,0.089,0.114,0.063
rdiff_flag,large_diff,large_diff,diff,similar,similar,diff,diff,large_diff,diff,similar,similar,diff,similar


### 2. 기술 통계량 표 저장

In [31]:
desc_df.to_excel("boston_qtcheck_desc.xlsx")

#### 💡 인사이트

**🔹 CRIM (범죄율)**
1. 최댓값과 중앙값 비율 13.1배 차이로 극한 분포의 큰편차
2. 상단 이상치 66개(13%)로 극단적인 고범죄 지역 존재
3. 왜도 5.22 + 첨도 37.13으로 극도의 오른쪽 꼬리 형태, log 변환 필수

**🔹 ZN (넓은 지역 비율)**
1. 최솟값이 0이어서 상대편차 무한대의 극한 경우
2. 상단 이상치 68개(13.4%)로 CRIM과 유사한 패턴
3. 왜도 2.23 + 뾰족한 분포로 log 변환 필요

**🔹 INDUS (산업 지역 비율)**
1. 상대편차 0.149로 중간 정도의 편차, 비교적 안정적
2. 이상치 0개 - 가장 균형잡힌 변수
3. 왜도 0.295(대칭) + 완만한 분포로 정상분포에 가까움

**🔹 NOX (산화질소 농도)**
1. 상대편차 0.031로 유사한 수준의 매우 안정적 분포
2. 이상치 0개로 완벽하게 정규분포
3. 왜도 0.729로 약간 오른쪽 치우침, 변환 불필요

**🔹 RM (평균 방 수)**
1. 상대편차 0.012로 극도로 유사한 균등분포
2. 상·하 이상치 균형있게 분포(상 4.3%, 하 1.6%)
3. 왜도 0.404 + 첨도 1.892로 중앙에 집중된 안정적 형태

**🔹 AGE (오래된 건물 비율)**
1. 상대편차 0.115로 중간 정도의 편차
2. 이상치 0개로 매우 안정적
3. 왜도 -0.599로 약간 왼쪽 치우침, 변환 불필요

**🔹 DIS (도심까지 거리)**
1. 상대편차 0.183으로 중간 정도의 편차
2. 상단 이상치 5개(1%)로 매우 경미함
3. 왜도 1.012 + 뾰족한 분포로 log 변환 필요

**🔹 RAD (고속도로 접근성)**
1. 상대편차 0.910으로 큰편차, 이산 값으로 인한 편차
2. 이상치 0개 - 범주형 성격의 변수
3. 왜도 1.005 + 완만한 분포로 log 변환 권장

**🔹 TAX (재산세)**
1. 상대편차 0.237로 중간 정도의 편차
2. 이상치 0개로 안정적
3. 왜도 0.670 + 완만한 분포로 거의 정상분포

**🔹 PTRATIO (학생-교사 비율)**
1. 상대편차 0.031로 유사한 수준, 매우 안정적
2. 하단 이상치만 15개(3%) - 소수 지역의 낮은 비율
3. 왜도 -0.802로 왼쪽 치우침, 변환 불필요

**🔹 B (흑인 인구 비율)**
1. 상대편차 0.089로 유사하나 의심스러운 데이터 분포
2. 하단 이상치 77개(15.2%) - 가장 높은 이상치 비율
3. 왜도 -2.890 + 첨도 7.227로 극도의 왼쪽 꼬리, 역 log 변환 필요

**🔹 LSTAT (낮은 사회경제적 지위)**
1. 상대편차 0.114로 중간 정도의 편차
2. 상단 이상치 7개(1.4%)로 경미함
3. 왜도 0.906 + 뾰족한 분포로 log 변환 필요

**🔹 MEDV (주택 중앙값, 타겟)**
1. 상대편차 0.063으로 유사, 안정적 분포
2. 상단 이상치 38개(7.5%) vs 하단 2개(0.4%) - 고가 주택 편향
3. 왜도 1.108 + 뾰족한 분포로 log 변환 권장